src/dlinkedlist/DLinkedList.java

```diff
@@ -0,0 +1,455 @@
+            if (!collection.contains(node.element)) {
+                result |= delete(node);
+            }
+            node = node.next;
+        }
+        return result;
+    }
+
+    @Override
+    public void clear() {
+        startup();
+    }
+
+    @Override
+    public E get(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<E> node = getNode(index);
+        assert node != null;
+        return node.element;
+    }
+
+    @Override
+    public E set(int index, E element) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<E> node = getNode(index);
+        assert node != null;
+        E elementToSet = node.element;
+        node.element = element;
+        return elementToSet;
+    }
+
+    @Override
+    public void add(int index, E element) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<E> node = getNode(index);
+        assert node != null;
+        insertBefore(node, element);
+    }
+
+    @Override
+    public E remove(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<E> node = getNode(index);
+        assert node != null;
+        E result = node.element;
+        delete(node);
+        return result;
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        int nodeIndex = 0;
+        Node<E> node = frontCapNode.next;
+        while (!node.equals(tailCapNode)) {
+            if (Objects.equals(node.element, o)) {
+                return nodeIndex;
+            }
+            nodeIndex++;
+            node = node.next;
+        }
+        return -1;
+    }
+
+    @Override
+    public int lastIndexOf(Object o) {
+        int nodeIndex = size - 1;
+        Node<E> node = tailCapNode.prev;
+        while (!node.equals(frontCapNode)) {
+            if (Objects.equals(node.element, o)) {
+                return nodeIndex;
+            }
+            nodeIndex--;
+            node = node.prev;
+        }
+
+        return -1;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        Node<E> node = frontCapNode.next;
+        while (!node.equals(tailCapNode)) {
+            if (!node.equals(frontCapNode.next)) {
+                stringBuilder.append(", ");
+            }
+            stringBuilder.append(node.element);
+            node = node.next;
+        }
+        return stringBuilder.toString();
+    }
+
+    @Override
+    public ListIterator<E> listIterator() {
+       return listIterator(0);
+    }
+
+
+    public ListIterator<E> listIterator(int index) {
+        return new ListIterator<E>() {
+
+            private final Node<E> helperNode = getNode(index);
+            private Node<E> current;
```

Comment:
Itt fölöslges a helperNode, mivel nem használod másra, csak hogy a current változó értékét beállítsd, ezt egy sorba össze lehet vonni.